pkg/scheduler/preemption/classical/hierarchical_preemption.go

```diff
@@ -0,0 +1,235 @@
+	for {
+		if hasHierarchicalAdvantage {
+			candidateList = &hierarchyCandidates
+		} else {
+			candidateList = &priorityCandidates
+		}
+		collectCandidatesInSubtree(ctx, trackingNode, trackingNode, previousRoot, hasHierarchicalAdvantage, candidateList)
+		if !trackingNode.HasParent() {
+			break
+		}
+		fits, remainingRequests = cache.WorkloadFitsInQuota(trackingNode, remainingRequests)
+		hasHierarchicalAdvantage = hasHierarchicalAdvantage || fits
+		previousRoot = trackingNode
+		trackingNode = trackingNode.Parent()
+	}
+	return hierarchyCandidates, priorityCandidates
+}
+
+// visit the nodes in the hierarchy and collect the ones that exceed quota
+// avoid subtrees that are within quota and the forbidden subtree
+func collectCandidatesInSubtree(ctx *HierarchicalPreemptionCtx, node *cache.CohortSnapshot, startingNode *cache.CohortSnapshot, forbiddenSubtree *cache.CohortSnapshot, hasHierarchicalAdvantage bool, result *[]*candidateElem) {
+	for _, childCohort := range node.ChildCohorts() {
+		if childCohort == forbiddenSubtree {
+			continue
+		}
+		// don't look for candidates in subtrees that are not exceeding their quotas
+		if cache.IsWithinNominalInResources(childCohort, ctx.FrsNeedPreemption) {
+			continue
+		}
+		collectCandidatesInSubtree(ctx, childCohort, startingNode, forbiddenSubtree, hasHierarchicalAdvantage, result)
+	}
+	for _, childCq := range node.ChildCQs() {
+		if childCq == ctx.Cq {
+			continue
+		}
+		if !cache.IsWithinNominalInResources(childCq, ctx.FrsNeedPreemption) {
+			*result = append(*result, getCandidatesFromCQ(childCq, startingNode, ctx, hasHierarchicalAdvantage)...)
+		}
+	}
+}
+
+// getNodeHeight calculates the distance to the furthest leaf
+func getNodeHeight(node *cache.CohortSnapshot) int {
+	maxHeight := min(len(node.ChildCQs()), 1)
+	for _, childCohort := range node.ChildCohorts() {
+		maxHeight = max(maxHeight, getNodeHeight(childCohort)+1)
+	}
+	return maxHeight
+}
+
+// FindHeightOfLowestSubtreeThatFits returns height of a lowest subtree in the cohort
+// that fits additional val of resource fr. If no such subtree exists, it returns
+// height the whole cohort hierarchy. Note that height of a trivial subtree
+// with only one node is 0. It also returns if the returned subtree is smaller than the whole cohort tree.
+func FindHeightOfLowestSubtreeThatFits(c *cache.ClusterQueueSnapshot, fr resources.FlavorResource, val int64) (int, bool) {
+	if !(c.BorrowingWith(fr, val) && c.HasParent()) {
+		return 0, c.HasParent()
+	}
+	remaining := cache.CalculateRemaining(c, fr, val)
+	var trackingNode *cache.CohortSnapshot
+	for trackingNode = c.Parent(); trackingNode.HasParent(); trackingNode = trackingNode.Parent() {
+		if trackingNode.ResourceNode.Usage[fr]+remaining <= trackingNode.ResourceNode.SubtreeQuota[fr] {
+			break
+		}
+		remaining = cache.CalculateRemaining(trackingNode, fr, remaining)
+	}
+	return getNodeHeight(trackingNode), trackingNode.HasParent()
```

Comment:
it is possible that we don't find a fit. Please document this behavior in docstring. Do we want to return the nodeHeight of root in this case, or what makes sense?
```suggestion
	// no fit found
	return getNodeHeight(c.Parent().Root()), false
```

Reply:
There is a comment above the function
```
If no such subtree exists, it returns height the whole cohort hierarchy.
```
Do we need more?

Reply:
Done.